packages/@aws-cdk/aws-stepfunctions/lib/state-machine.ts

```diff
@@ -148,6 +148,87 @@ abstract class StateMachineBase extends Resource implements IStateMachine {
       resourceArns: [this.stateMachineArn],
     });
   }
+
+  /**
+   * Grant the given identity permissions to read results from state
+   * machine.
+   */
+  public grantRead(identity: iam.IGrantable): iam.Grant {
+    iam.Grant.addToPrincipal({
+      grantee: identity,
+      actions: [
+        'states:ListExecutions',
+        'states:ListStateMachines',
+      ],
+      resourceArns: [this.stateMachineArn],
+    });
+    iam.Grant.addToPrincipal({
+      grantee: identity,
+      actions: [
+        'states:DescribeExecution',
+        'states:DescribeStateMachineForExecution',
+        'states:GetExecutionHistory',
+      ],
+      resourceArns: [`${this.executionArn()}:*`],
+    });
+    return iam.Grant.addToPrincipal({
+      grantee: identity,
+      actions: [
+        'states:ListActivities',
+        'states:DescribeStateMachine',
+        'states:DescribeActivity',
+      ],
+      resourceArns: ['*'],
+    });
+  }
+
+  /**
+   * Grant the given identity task response permissions for a resource
+   *
+   * @default stateMachineArn
+   */
+  public grantTaskResponse(identity: iam.IGrantable, activityArn?: string): iam.Grant {
+    // Validate activityArn
+    if (activityArn && Arn.parse(activityArn, ':').resource !== 'activity') {
+      throw new Error('activityArn must be a valid activity ARN');
+    }
+
+    // ActivityArn is valid if it exists
+    const arn = activityArn || this.stateMachineArn;
+
+    return iam.Grant.addToPrincipal({
+      grantee: identity,
+      actions: [
+        'states:SendTaskSuccess',
+        'states:SendTaskFailure',
+        'states:SendTaskHeartbeat',
+      ],
+      resourceArns: [arn],
+    });
+  }
+
+  /**
+   * Grant the given identity custom permissions
+   */
+  public grant(identity: iam.IGrantable, actions: string[], resourceArn: string): iam.Grant {
```

Comment:
I don't believe any of our other grant APIs allow specifying a resource explicitly (i.e signature of `grant(identity: iam.IGrantable, actions: string[])`.

Since we have multiple resources that this can be scoped down to, we should try to determine the intent and ask for the information required rather than simply pass through an entire ARN. I'm not sure how it will work when an arn for another account or region is supplied. I don't believe Step Functions will support the x-region boundary and x-account is problematic.


what happens when a `resourceArn` that has nothing to do with Step Functions is supplied?? will we issue permissions?

Can we default to the state machine Arn and make it optional? Simply calling grant with a principal and some actions should default to the most sensible thing.

Reply:
Default to state machine arn makes sense, so I'll make the `resourceArn` parameter optional. As for when the user specifies the wrong resourceArn, right now the method will issue permissions but they will not work. Is that problematic? 

I envision the user might need to specify execution ARNs or activity ARNs inside `grant()`. Apart from ensuring that the resource is an `activity` or `execution`, I don't know how we might validate the specified resourceArn. Is it necessary to add this additional check inside of `grant()`?

Reply:
@shivlaks and I spoke offline and agreed to do the heavy lifting behind the scenes and not ask the user for an ARN at all.